src/Factory/ServerRequestFactory.php

```diff
@@ -4,35 +4,178 @@
 
 namespace Nyholm\Psr7\Factory;
 
+use InvalidArgumentException;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Message\UploadedFileInterface;
 use Interop\Http\Factory\ServerRequestFactoryInterface;
 use Nyholm\Psr7\ServerRequest;
+use Nyholm\Psr7\Factory\UriFactory;
+use Nyholm\Psr7\UploadedFile;
 
 /**
  * @author Tobias Nyholm <tobias.nyholm@gmail.com>
  * @author Martijn van der Ven <martijn@vanderven.se>
  */
 class ServerRequestFactory implements ServerRequestFactoryInterface
 {
-    public function createServerRequest($method, $uri)
+    public function createServerRequest($method, $uri): ServerRequestInterface
     {
         return new ServerRequest($method, $uri);
     }
 
-    public function createServerRequestFromArray(array $server)
+    public function createServerRequestFromArray(array $server): ServerRequestInterface
     {
         if (!isset($server['REQUEST_METHOD'])) {
-            throw new \InvalidArgumentException('Cannot determine HTTP method');
+            throw new InvalidArgumentException('Cannot determine HTTP method');
         }
-        // TODO: find a MUCH better way
         $method = $server['REQUEST_METHOD'];
-        $SERVER = $_SERVER;
-        $_SERVER = $server;
-        // Until https://github.com/guzzle/psr7/pull/116 is resolved
-        if (!isset($_SERVER['HTTPS'])) {
-            $_SERVER['HTTPS'] = 'off';
+
+        $uri = (new UriFactory())->createUriFromArray($server);
+        if ($uri->getScheme() === '') {
+            $uri = $uri->withScheme('http');
+        }
+
+        $protocol = isset($server['SERVER_PROTOCOL']) ? str_replace('HTTP/', '', $server['SERVER_PROTOCOL']) : '1.1';
+
+        return new ServerRequest($method, $uri, [], null, $protocol, $server);
+    }
+
+    /**
+     * Create a new server request from a set of arrays.
+     *
+     * @param array $server Typically $_SERVER or similar structure.
+     * @param array $cookie Typically $_COOKIE or similar structure.
+     * @param array $get    Typically $_GET or similar structure.
+     * @param array $post   Typically $_POST or similar structure.
+     * @param array $files  Typically $_FILES or similar structure.
+     *
+     * @throws InvalidArgumentException If no valid method or URI can be determined.
+     *
+     * @return ServerRequestInterface
+     */
+    public function createServerRequestFromArrays(
+        array $server,
+        array $cookie,
+        array $get,
+        array $post,
+        array $files
+    ): ServerRequestInterface {
```

Comment:
The first 12 lines of `createServerRequestFromArrays` are an exact copy of `createServerRequestFromArray`, just so they are equalised. What would be the best way to reduce this duplicate code?

I first thought I would do `$this->createServerRequestFromArray($server)` on the first line, but that makes adding the headers harder.